atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraServersConfigs.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+
+import java.net.InetSocketAddress;
+import java.util.Set;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.collect.ImmutableSet;
+
+public final class CassandraServersConfigs {
+    protected CassandraServersConfigs() {
+
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type",
+            defaultImpl = DefaultCassandraServersCqlDisabledConfig.class)
+    @JsonSubTypes(
+            {
+                    @JsonSubTypes.Type(value = DefaultCassandraServersCqlDisabledConfig.class,
+                            name = DefaultCassandraServersCqlDisabledConfig.TYPE),
+                    @JsonSubTypes.Type(value = ImmutableExplicitCassandraServersCqlDisabledConfig.class,
+                            name = ExplicitCassandraServersCqlDisabledConfig.TYPE),
+                    @JsonSubTypes.Type(value = ImmutableCassandraServersCqlEnabledConfig.class,
+                            name = CassandraServersCqlEnabledConfig.TYPE)
+            }
+    )
+    public interface CassandraServersConfig {
+        Set<InetSocketAddress> thrift();
+        Set<InetSocketAddress> cql();
+        String type();
+    }
+
+    @JsonDeserialize(as = DefaultCassandraServersCqlDisabledConfig.class)
+    @JsonTypeName(DefaultCassandraServersCqlDisabledConfig.TYPE)
+    public static class DefaultCassandraServersCqlDisabledConfig implements CassandraServersConfig {
+        public static final String TYPE = "default";
+
+        private Set<InetSocketAddress> thriftServers;
+
+        @JsonCreator
+        public DefaultCassandraServersCqlDisabledConfig(InetSocketAddress... thriftServers) {
+            this.thriftServers = ImmutableSet.copyOf(thriftServers);
+        }
+
+        @JsonCreator
+        public DefaultCassandraServersCqlDisabledConfig(Iterable<InetSocketAddress> thriftServers) {
+            this.thriftServers = ImmutableSet.copyOf(thriftServers);
+        }
+
+        @Override
+        @JsonValue
+        public Set<InetSocketAddress> thrift() {
+            return thriftServers;
+        }
+
+        @Override
+        public Set<InetSocketAddress> cql() {
+            return ImmutableSet.of();
```

Comment:
you  *could* throw an exception, or consider the possibility of using a visitor, still working out the ergonomics in my head but I imagine something along the lines of:

```java
createClientPool(config) {
  config.visit(new Visitor<ClientPool>() {
    
    @Override
    public ClientPool thriftOnly(List<InetSocketAddress> bla) {
      return ThriftClientPool.create(...);
    }

    @Override
    public ClientPool cqlCapable(List<CassandraPorts> cassPorts, AsyncConfig config) {
      // or you can pass in the config itself if you want to add a further flag/gate 
      // for whether or not to use async despite it being specified
      if (config.isEnabled()) {
        return AsyncClientPool.create(...);
      } else {
        return thriftOnly(cassPorts.stream().map(CassandraPorts::thriftAddress).collect(toList()));
      }
    }
  });
}
```
The advantage of this approach is that it removes ambiguity/somewhat ill defined semantics for an empty list of cql ports.


Above/below is pseudo-code that almost compiles, but effectively you have a config interface that looks like this:

```java
public interface CassandraServersConfig {
  Set<InetSocketAddress> thrift();

  <T> T visit(Visitor<T> visitor);

  public interface Visitor<T> {
     T thriftOnly(List<InetSocketAddress> thriftAddresses);
     T cqlCapabable(List<CassandraPorts> cqlPlusThriftPorts);
  }
}
```

where for example the `DefaultDisabledConfig` looks something like this:

```java
public abstract class DefaultDisabledConfig extends CassandraServersConfig {
  public abstract Set<InetSocketAddress> thrift();
  
  public final <T> T visit(Visitor<T> visitor) {
    return visitor.visit(thrift());
  }
  
  ...
}
```

etc. etc.

Reply:
Visitor seems like an elegant solution for this I'll explore it. I have some concerns about declaring the interface for visitors in `CassandraServersConfig` since the ClientPool, etc. depends on the whole config.

Reply:
Implemented one approach for `Visitor` should be double checked if it should simpler.